### ISSUE TITLE
cmd/compile: boolean operations not constant folded

### ISSUE THREAD

**Issue by marigonzes** [golang/go#32713]
### What version of Go are you using (`go version`)?
<pre>
$ go version
go version devel +c82e7e7 Tue Jun 18 22:21:45 2019 +0000 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes.

### What did you do?
I inspected the assembly code generated for the following functions:
```go
func f1(b bool) bool {
    return b && !b
}
```

```go
func f2(b bool) bool {
    return b || !b
}
```

### What did you expect to see?
I expected them to be compiled to a simple return: false in the first case, true in the second.

### What did you see instead?
Instead, boolean operations are performed on the variable `b`.

**Comment by martisch** [comment]
Where those examples constructed or found in real code? Im not sure how often they would trigger.

What is the scope of the feature request: to support these specific cases or be able to prove all tautologies and contradictions in boolean expressions (e.g. pierce law ((P→Q)→P)→P written as: !(!p || q) || p)?

Im asking because if its the former we want to collect cases that are easy to detect and happen frequently and if its the later whether go should contain a full propositional prover (e.g. SAT solver) applied to boolean expressions that also keeps side effects if any.

**Comment by marigonzes** [comment]
I found these examples out of curiosity, because I wanted to know how the Go compile would behave, but I imagine code like this being auto-generated.
I posted this issue just to know if anyone was thinking about this.
But if code like this doesn't appear in the wild, I think we should close it.

**Comment by cherrymui** [comment]
> Interestingly gcc Go seems to not completely optimize p || true: https://godbolt.org/z/peYHWU

You need to specify optimization level. The default is -O0, which does not optimize. -O1 and above do.

**Comment by cherrymui** [comment]
As @martisch mentioned, `b || !b` is not constant folding, as the input is not constant. It is true that for all possible input, the expression is false, but normal constant folding wouldn't do that. If this doesn't come up in real code, I agree that it is not worth the extra complexity.

**Comment by martisch** [comment]
> You need to specify optimization level. The default is -O0, which does not optimize. -O1 and above do.
Oops. Thank you for the correction. I forget this sometimes as Go thankfully does optimization without added knobs.

**Comment by martisch** [comment]
> You need to specify optimization level. The default is -O0, which does not optimize. -O1 and above do.
Oops. Thanks for the correction. I was distracted from adding flags by GCC C++ seemingly optimizing this even without any additional flags in godbolt: https://godbolt.org/z/SMYxsz

**Comment by marigonzes** [comment]
Personally, I've never seen code like this. I was just curious about the output of the compiler, given this two functions.
Leaving the decision to you :wink:

**Comment by katiehockman** [comment]
Since this seems like an edge case that won't appear in the wild or cause people any issues, it sounds like we should close this. @martisch can you confirm/close this if so?

**Comment by martisch** [comment]
TL;DR I think we should wait to see an actual use case where this would help and then implement some optimizations for propositional tautologies/contradicions preferably in the compiler backend.

This might very well come up after some optimizations triggered in the compiler backend but then detecting tautologies/contradictions in the frontend for simple cases like shown wont help. To not solve the made up cases and not detect the actual cases where this triggers I would think we should wait for code in the wild to show up that could use this. If somebody has an example that can not be just simplified by code review without compromising readability I think we can close this.

Linters and static code checks meanwhile can detect simple cases such as `b || q || !b` and suggest simplification.
